### ISSUE TITLE
Invalid Program Exception Sometimes

### ISSUE THREAD

**Issue by curtiswellborn** [elastic/elasticsearch-net#5335]
**Common Language Runtime detected an invalid at [Elasticsearch.Net]**:
This is randomly thrown from two of our microservices with Nest installed. A fresh build/deploy fixes the issue for some period of time. 

**Open discussion on ES forum(not going anywhere)**
https://discuss.elastic.co/t/nest-6-8-3-getting-invalidprogramexception-sometimes/227037

**Envionrment:**
AWS ECS Fargate
Docker container
Build/Test - linux - alpine(mcr.microsoft.com/dotnet/core/sdk:3.1-alpine)
Container - linux - debian(mcr.microsoft.com/dotnet/core/aspnet:3.1)
.Net Core 3.1
NEST 6.8.6

Stacktrace:
```
Message Common Language Runtime detected an invalid program.stackTrace at [Elasticsearch.Net](http://elasticsearch.net/).Transport `1.RequestAsync[TResponse](HttpMethod method, String path, CancellationToken cancellationToken, PostData data, IRequestParameters requestParameters) at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) at [Elasticsearch.Net](http://elasticsearch.net/).Transport` 1.RequestAsync[TResponse](HttpMethod method, String path, CancellationToken cancellationToken, PostData data, IRequestParameters requestParameters)
at [Elasticsearch.Net](http://elasticsearch.net/).ElasticLowLevelClient.DoRequestAsync[TResponse](HttpMethod method, String path, CancellationToken cancellationToken, PostData data, IRequestParameters requestParameters)
at [Elasticsearch.Net](http://elasticsearch.net/).ElasticLowLevelClient.SearchAsync[TResponse](String index, String type, PostData body, SearchRequestParameters requestParameters, CancellationToken ctx)
at Nest.LowLevelDispatch.SearchDispatchAsync[TResponse](IRequest `1 p, SerializableData` 1 body, CancellationToken ct)
at Nest.ElasticClient.b__563_0[T,TResult](ISearchRequest p, SerializableData `1 d, CancellationToken c) at Nest.ElasticClient.Nest.IHighLevelToLowLevelDispatcher.DispatchAsync[TRequest,TQueryString,TResponse,TResponseInterface](TRequest request, CancellationToken cancellationToken, Func` 3 responseGenerator, Func`4 dispatch)
at App.Core.Search.SearchRepository.SearchAsync(BenchmarkSearchQuery search, CancellationToken cancellationToken) in /src/App/Core/Search/SearchRepository.cs:line 39
at App.Controllers.SearchController.SearchAsync(BenchmarkSearchRequest request, CancellationToken cancellationToken) in /src/App/Controllers/SearchController.cs:line 77
at lambda_method(Closure , Object )
at Microsoft.Extensions.Internal.ObjectMethodExecutorAwaitable.Awaiter.GetResult()
at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.AwaitableObjectResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object arguments)
at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.g__Logged|12_1(ControllerActionInvoker invoker)
at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()
--- End of stack trace from previous location where exception was thrown ---
at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.g__Awaited|25_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
```

**Solution Attempts:**
*Failed* - We have tried moving the Nest version
*Pending* - Have the sdk image and runtime image on the same flavor of linux. This one, imo, would be very odd. I should be able to build my solution anywhere and it run on any OS. 
*Short Term Success* - Kick off a jenkins job to build a new image and redeploy it. This is reactive but works for when we encouner the issue. It does lead to downtime for our clients.

**Comment by stevejgordon** [comment]
Hi @curtiswellborn. Thanks for reporting this issue. I suspect it is going to be a hard one to isolate the cause.

I had the same initial opinion as Russ who responded to your discussion. The most common cause for such an exception is badly formed IL, generated at runtime. Based on the stack trace and where this is occurring, it's at the point that is creating/sending the request. Given that 6.x uses Newtonsoft.Json, that's one possible source of the issue during serialization. However, like Russ, I'd expect a problem in that library to have been reported from many other sources.

We have no IL emit code in 6.x so it seems unlikely it's directly the client code triggering this. Without a dump from the crash, it'll be hard to reconstruct or replicate this.

While I don't want to unfairly point the blame elsewhere, I do have to suspect dependencies that use IL emit APIs as the most likely source of this issue. From your original discussion, you mentioned you are using the DataDog APM. That does have a [few issues logged](https://github.com/DataDog/dd-trace-dotnet/search?q=detected+an+invalid+program&type=issues) which report similar exceptions. [This issue](https://github.com/DataDog/dd-trace-dotnet/issues/762) seems most pertinent as it lists a fix for this exception in 1.19.2. Please could you try upgrading your DataDog APM dependency to the latest release as the first course of action? If that continues to fail, could you run this service without DataDog for a while to see if that resolves the issue? Given that APM will instrument requests, it's possible that your application is occasionally hitting this problem.

**Comment by curtiswellborn** [comment]
Hey @stevejgordon Thanks for the quick response. Our conversations with Datadog in the past didn't lead to anything. Looks like they did figure something out. Thank you for the find. We will attempt to update to the latest version, 1.16.2 to 1.24.0, of the tracer and see if that resolves the issue.
